AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/test/Keyrings/AwsKms/AwsKmsHierarchicalKeyring/TestAwsKmsHierarchicalKeyring.dfy

```diff
@@ -320,6 +320,471 @@ module TestAwsKmsHierarchicalKeyring {
+
+    // Initialize the shared Cryptographic Materials Cache
+    var sharedCacheInput :- expect mpl.CreateCryptographicMaterialsCache(
+      Types.CreateCryptographicMaterialsCacheInput(
+        cache := Types.CacheType.Default(
+          Types.DefaultCache(
+            entryCapacity := 100
+          )
+        )
+      )
+    );
+
+    // Wrap the initialized shared Cryptographic Materials Cache as a `Shared` CacheType object
+    var sharedCache := Types.CacheType.Shared(
+      sharedCacheInput
+    );
+
+    // Create a Hierarchy Keyring HK1 with branchKeyIdWest,
+    // keyStoreClientRegionWest and the shared cache `sharedCache`
+    // without specifying the partitionId, so that it is initialized to a random UUID
+    var hierarchyKeyring1 :- expect mpl.CreateAwsKmsHierarchicalKeyring(
+      Types.CreateAwsKmsHierarchicalKeyringInput(
+        branchKeyId := Some(branchKeyIdWest),
+        branchKeyIdSupplier := None,
+        keyStore := keyStoreClientRegionWest,
+        ttlSeconds := ttl,
+        cache := Some(sharedCache)
+      )
+    );
+
+    // Create a Hierarchy Keyring HK2 with branchKeyIdWest,
+    // keyStoreClientRegionEast and the shared cache `sharedCache`
+    // without specifying the partitionId, so that it is initialized to a random UUID
+    var hierarchyKeyring2 :- expect mpl.CreateAwsKmsHierarchicalKeyring(
+      Types.CreateAwsKmsHierarchicalKeyringInput(
+        branchKeyId := Some(branchKeyIdWest),
+        branchKeyIdSupplier := None,
+        keyStore := keyStoreClientRegionEast,
+        ttlSeconds := ttl,
+        cache := Some(sharedCache)
+      )
+    );
+
+    // Get test materials
+    var materials := GetTestMaterials(TEST_ESDK_ALG_SUITE_ID);
+
+    // Try encrypting the test materials with HK2, which has branchKeyIdWest
+    // but its keystore is keyStoreClientRegionEast
+    var encryptionMaterialsOutMismatchedRegion := hierarchyKeyring2.OnEncrypt(
+      Types.OnEncryptInput(materials:=materials)
+    );
+
+    // This encryption should fail because of region mismatch
+    expect encryptionMaterialsOutMismatchedRegion.IsFailure();
+
+    // Encrypt and Decrypt round trip for the test materials with HK1,
+    // which has branchKeyIdWest and its keystore is keyStoreClientRegionWest
+    // This should pass
+    TestRoundtrip(hierarchyKeyring1, materials, TEST_ESDK_ALG_SUITE_ID, branchKeyIdWest);
+
+    // Again, try encrypting the test materials with HK2, which has branchKeyIdWest
+    // but its keystore is keyStoreClientRegionEast
+    var encryptionMaterialsOutMismatchedRegionFromCache := hierarchyKeyring2.OnEncrypt(
+      Types.OnEncryptInput(materials:=materials)
+    );
+
+    // This encryption should fail because the partition IDs for HK1 and HK2 are different
+    // even though the partition IDs are unspecified. In such a case, partition IDs are
+    // initialized as UUIDs which have negligible probability of collision
```

Comment:
```suggestion
    // This encryption should fail because the partition IDs for HK1 and HK2 are different
    // even though the partition IDs are unspecified. In such a case, partition IDs are
    // initialized as UUIDs which have negligible probability of collision.
    // This proves the cache was missed.
```

Reply:
Thanks